/**
 * Copyright Soramitsu Co., Ltd. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

syntax = "proto3";

package iroha.protocol;

import "transaction.proto";
import "queries.proto";
import "qry_responses.proto";
import "google/protobuf/empty.proto";

enum TxStatus {
  STATELESS_VALIDATION_FAILED = 0;
  STATELESS_VALIDATION_SUCCESS = 1;
  STATEFUL_VALIDATION_FAILED = 2;
  STATEFUL_VALIDATION_SUCCESS = 3;
  COMMITTED = 4;
  MST_EXPIRED = 5;
  NOT_RECEIVED = 6;
}

message ToriiResponse {
  TxStatus tx_status = 1;
  bytes tx_hash = 2;
  string error_message = 3;
}

message TxStatusRequest{
  bytes tx_hash = 1;
}

message TxList {
  repeated Transaction transactions = 1;
}

service CommandService {
  rpc Torii (Transaction) returns (google.protobuf.Empty);
  rpc ListTorii (TxList) returns (google.protobuf.Empty);
  rpc Status (TxStatusRequest) returns (ToriiResponse);
  rpc StatusStream(TxStatusRequest) returns (stream ToriiResponse);
}

service QueryService {
  rpc Find (Query) returns (QueryResponse);
  rpc FetchCommits (BlocksQuery) returns (stream BlockQueryResponse);
}
